### PR TITLE
tests: getRestartDelay waits for the next Terminated state 

### LINKED ISSUE #71949: tests: MaxContainerBackOff tests are flaky when using Windows nodes
**Which jobs are failing**:

N/A

**Which test(s) are failing**:

    [k8s.io] Pods  should cap back-off at MaxContainerBackOff [Slow][NodeConformance]
    [k8s.io] Pods  should have their auto-restart back-off timer reset on image update [Slow][NodeConformance]

**Since when has it been failing**:

Unknown. Tests are flaky when using Windows nodes.

**Testgrid link**:

N/A

**Reason for failure**:

When using Windows nodes, the mentioned tests might fail randomly due to the fact that Windows Kubelet might fail to report the new Pod Status before its execution finishes, causing the Pod Status to remain in ``CrashLoopBackOff`` state, which causes the tests to fail, since they're expecting the Pod to enter a ``Running`` State.

This could be solved by increasing the Pod's sleep period from 5 second to 10 seconds, allowing sufficient time for Kubelet to report the new ``Running`` state, or, instead of waiting for the ``Running`` state to occur, the tests should wait for the ``RestartCount`` to increase.

**Anything else we need to know**:

/sig windows

/kind failing-test

### PR DESCRIPTION
**Pull request by claudiubelu**
**What type of PR is this?**

/kind failing-test
/kind flake

/sig testing
/sig windows

**What this PR does / why we need it**:

Kubelet might miss reporting the new Running state when restarting
a pod after its backoff period expired, and thus, the pod will
continue to remain in CrashLoopBackOff state, causing the
"should cap back-off at MaxContainerBackOff" and
"should have their auto-restart back-off timer reset on image update"
tests to fail, since they're waiting the Pods to enter a Running state.

Waiting for the next Terminated state instead of the next Running state
is more reliable.

Note that this adds 5 seconds to the restart delay due to the fact that
the Container runs for 5 seconds (it's command is "sleep 5"), but it is
within the test's expectations.

**Which issue(s) this PR fixes**:

Fixes #71949

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by yujuhong — CHANGES_REQUESTED** [review]
Can we get a better understanding of why this is slower on Windows, before bumping the timeout?

**Comment by benmoss** [comment]
I have tried reproducing this problem on my cluster and haven't seen it fail. How often do you all see this failure?

**Comment by oomichi** [comment]
The reason why this bumps the timeout to 10sec is the sleep time is 10 sec as syncLoopFrequency before waiting Running status of the pod like https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/pods.go#L647-L648 , right?
I also am not sure this yet, it is better to describe the e2e operation log which includes
```
"getRestartDelay: restartCount = %d, finishedAt=%s restartedAt=%s (%s)"
```
on the issue report or this change message to share that.

**Comment by oomichi** [comment]
cc @oomichi

**Comment by nikopen** [comment]
@bclau  @yujuhong any chance this is merged in the following 2 weeks? 
Code Freeze is in effect from next Friday.

**Comment by bclau** [comment]
So, from what I've seen, this only happens when using azure-vnet-cni, including the latest release v1.0.18-rc1. This problem does not occur when using flannel l2bridge or overlay.

So, basically what happens is this:

- pod sandbox is setup.
- container image is pulled,
- container is created.
- container is **started**.
- the CNI plugin is called to setup the container's networking (HotAttaching the endpoint to the running container)
- Finally, after all that is done, the Container **should** be put in the Running state. But, because the previous step can take a few seconds, the Container finishes its execution before the endpoint is attached, so the state transition is ContainerStarting -> Terminated instead of ContainerStarting -> Running.

The test explicitly expects the the Container to enter a Running state and that the RestartCount to be incremented. IMO, just checking if the RestartCount increased should be sufficient. If it is not sufficient, this sleep period increase is long enough for the CNI operation to finish and the ContainerStarting -> Running state transition to occur.

**Comment by benmoss** [comment]
Why does it only happen on the Azure CNI?

**Comment by spiffxp** [comment]
>This sleep period increase is long enough for the CNI operation to finish and the ContainerStarting -> Running state transition to occur.

I feel like we're not happy with the idea of introducing a sleep just to get a test to pass.  Here lies the path to unending tests and more flakes.

>IMO, just checking if the RestartCount increased should be sufficient. 

This sounds like it could be better, what would this change look like?

**Comment by yujuhong** [comment]
> * Finally, after all that is done, the Container **should** be put in the **Running** state. But, because the previous step can take a few seconds, the Container finishes its execution before the endpoint is attached, so the state transition is ContainerStarting -> Terminated instead of ContainerStarting -> Running.

@bclau the container you mentioned here, is it the regular user container or the pod sandbox? 

I don't understand fully why container would go into the terminated state *and then* became running. Could you explain a little bit more on what caused the state transition for the container?

**Comment by bclau** [comment]
> 
> > IMO, just checking if the RestartCount increased should be sufficient.
> 
> This sounds like it could be better, what would this change look like?

I amended the commit. Now getRestartDelay waits for the next Terminated state instead of the next Running state (which can be missed). It is still a simple change and it makes it more reliable. See test run: https://paste.ubuntu.com/p/d2PckYv3BR/

> 
> 
> > * Finally, after all that is done, the Container **should** be put in the **Running** state. But, because the previous step can take a few seconds, the Container finishes its execution before the endpoint is attached, so the state transition is ContainerStarting -> Terminated instead of ContainerStarting -> Running.
> 
> @bclau the container you mentioned here, is it the regular user container or the pod sandbox?
> 
> I don't understand fully why container would go into the terminated state _and then_ became running. Could you explain a little bit more on what caused the state transition for the container?

The user container.

I never said it enters the (Kubernetes) ``Running`` state after entering the (Kubernetes) ``Terminated`` state.

There is a **distinction** between the container being started and running and the reported ``Running`` pod state in Kubernetes and they should not be confused with each other, they are not the same. I have already explained the steps that occur when spawning a new pod.

Basically, if the Container's command finishes fast enough, it might terminate before the Kubernetes ``Running`` state could be observed.

And just an FYI, because of this issue, the test is currently flaky on gce-windows: https://testgrid.k8s.io/sig-windows#gce-windows-master

**Comment by yujuhong** [comment]
The current fix looks reasonable.
/lgtm

**Review by yujuhong — APPROVED** [review]

**Comment by spiffxp** [comment]
/approve
I was waiting for @yujuhong's take on this, thanks for your patience all

**Comment by michmike** [comment]
nice fix @bclau
